### ISSUE TITLE
Linking the GC

### ISSUE THREAD

**Issue by jacob-hughes** [softdevteam/rustcgc#4]
Having the GC runtime link correctly to the runtime of user programs is proving quite difficult. At a high-level, there are 3 requirements that we need to satisfy:

1. Parts of its API must be called from within Rust's `libstd`. For example, suppose we want to initialise our GC heap at the start of execution of our user's GC'd bin: we might want to make a call to something like `gc::init()` within `rustcgc/src/libstd/rt.rs::22` (where the setup code before calling the user's main method is executed).

2. We want to expose parts of the GC to the user program too, so that we can call something like `gc::force()` to invoke a collection for debug purposes.

3. Safepoint polls are not inserted into user code until we hit LLVM. This is done with a pass which inserts `gc.safepoint_poll()` calls to the GC runtime at various points in the generated IR. These polls ask the GC "Do I need to GC, because I'm at a safepoint". My mental model of this is that _if_ the original user's binary is linked to the GC runtime, and the runtime implements the `gc.safepoint_poll()` function interface, this should just work.

With this in mind. I need something like this:

```
 +----------+
 |          |
 |  libstd  +--------------------+
 |          |                    |
 +-------+--+                    |
         ^                       |
         |                       |
         |                       |
         |                       |
         |                       v
+------------------+         +---+---+
|                  |         |       |
|  user's program  +-------->+  gc   |
|                  |         |       |
+------------------+         +-------+
```

**Approach 1: link gc runtime to (rustcgc compiled) user exe**
---
I'd been trying this for a while, but it falls apart when you need any kind of language level interaction with the GC. Realistically, Rust's `libstd` needs compiling with some hooks into the GC runtime.

**Approach 2: the gc runtime lives inside `libstd`**
---
Probably the most straightforward approach. It solves a couple of really annoying problems for us: the compiler takes care of all linking; we can use a single repo; and using the GC is really straightforward from the end-user's perspective. The deal breaker with this, however, is that building anything inside `libstd` is painfully slow. We can't really make effective use of the incremental compilation stuff and the development time between compiles would just make experimenting infeasible.

**Approach 3: use FFIs and the `#[link(..)]` attribute as GC API glue**
---
This is IMO our most realistic approach. I'm hoping (and running experiments that will either confirm or deny) that this will allow us to link `libstd` to our GC runtime and allow us to modify the GC without having to rebuild `libstd`. What I'm hoping will happen is `x.py build` will re-run the linking phase but notice that it doesn't need to actually re-compile anything. In addition, we will use an `asm` shim to pass the stack pointer to our GC's stack walk function so this approach would need to be taken in some form anyway.

---
**tl;dr** 
It looks like approach 3 (linking to the GC through an API of FFI calls in libstd) seems the most sensible. Do either of you have any thoughts about this? Does this sound sensible?

**Comment by ltratt** [comment]
I'm fine with approach 3 *but* it's not magic pixie dust. *Some* (maybe even most) changes will not require a full recompile, but some will.

There is also something else to consider: putting stuff into an FFI can make it more awkward to deal with types. It's not a deal breaker, but it might again lead to less stuff being put in the external lib than one first imagines.

**Comment by jacob-hughes** [comment]
I don't think there's a perfect solution here. I imagine once the GC becomes more mature it might be worth considering moving it into `libstd`.

> putting stuff into an FFI can make it more awkward to deal with types.

I'm hoping that this won't be as much of an issue as it sounds. At present, the GC runtime will expose a couple of functions as a minimal API: `force_gc()`, `init()`, `safepoint_poll()`. None of these take any arguments (`force_gc` and  `safepoint_poll` will, however, obtain the SP from some assembly glue).
